ngraph/core/src/op/dft.cpp

```diff
@@ -0,0 +1,196 @@
+    : Op({data, axes, signal_size})
+{
+    constructor_validate_and_infer_types();
+}
+
+bool op::v7::DFT::visit_attributes(AttributeVisitor& visitor)
+{
+    NGRAPH_OP_SCOPE(v7_DFT_visit_attributes);
+    return true;
+}
+
+std::shared_ptr<Node> op::v7::DFT::clone_with_new_inputs(const OutputVector& new_args) const
+{
+    NGRAPH_OP_SCOPE(v7_DFT_clone_with_new_inputs);
+    check_new_args_count(this, new_args);
+    NODE_VALIDATION_CHECK(
+        this, new_args.size() == 2 || new_args.size() == 3, "Number of inputs must be 2 or 3");
+
+    if (new_args.size() == 2)
+    {
+        return std::make_shared<op::v7::DFT>(new_args.at(0), new_args.at(1));
+    }
+
+    return std::make_shared<op::v7::DFT>(new_args.at(0), new_args.at(1), new_args.at(2));
+}
+
+void op::v7::DFT::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v7_DFT_validate_and_infer_types);
+
+    size_t num_of_inputs = get_input_size();
+
+    NODE_VALIDATION_CHECK(
+        this, num_of_inputs == 2 || num_of_inputs == 3, "DFT must have 2 or 3 inputs.");
+
+    element::Type input_et = get_input_element_type(0);
+    NODE_VALIDATION_CHECK(this,
+                          input_et == element::f32 || input_et == element::f16 ||
+                              input_et == element::bf16,
+                          "Input element type must be f32, f16, or bf16");
+
+    element::Type axes_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          axes_et == element::i64 || axes_et == element::i32 ||
+                              axes_et == element::u32 || axes_et == element::u64,
+                          "Axes element type must be i32, i64, u32 or u64");
+
+    PartialShape input_shape = PartialShape(get_input_partial_shape(0));
+    PartialShape axes_shape = PartialShape(get_input_partial_shape(1));
+    PartialShape output_shape = input_shape;
+
+    if (input_shape.rank().is_dynamic() || axes_shape.rank().is_dynamic() ||
+        !is_type<op::Constant>(input_value(1).get_node()))
+    {
+        set_output_type(0, get_input_element_type(0), output_shape);
+        return;
+    }
+
+    const auto input_rank = input_shape.rank().get_length();
+
+    NODE_VALIDATION_CHECK(this,
+                          axes_shape.rank().get_length() == 1,
+                          "Axes input must be 1D tensor. Got axes input rank: ",
+                          axes_shape.rank().get_length());
+
+    auto last_dim_with_two = input_shape[input_rank - 1] & Dimension(2);
+    NODE_VALIDATION_CHECK(this,
+                          !last_dim_with_two.get_interval().empty(),
+                          "The last dimension of input data must be 2. Got: ",
+                          input_shape[input_rank - 1]);
+
+    const auto& const_axes = get_constant_from_source(input_value(1));
+    auto axes = const_axes->cast_vector<int64_t>();
+
+    for (int64_t& axis : axes)
+    {
+        if (axis < 0)
+        {
+            axis += input_rank - 1;
```

Comment:
Probably here should be axis += input_rank?

Reply:
No, here should be `axis += input_rank - 1`, because, according to the operation specification, `-1` corresponds to the dimension `input_rank - 2`, the last input tensor dimension should be equal to `2`, and axis `input_rank - 1` is not admissible:

> *   **2**: `axes` - 1D tensor of type *T_IND* specifying dimension indices where DFT is applied, and `axes` is any unordered list of indices of different dimensions of input tensor, for example, `[0, 4]`, `[4, 0]`, `[4, 2, 1]`, `[1, 2, 3]`, `[-3, 0, -2]`. These indices should be integers from `-(r - 1)` to `(r - 2)` inclusively, where `r = rank(data)`. A negative axis `a` is interpreted as an axis `r - 1 + a`. Other dimensions do not change. The order of elements in `axes` attribute matters, and is mapped directly to elements in the third input `signal_size`. Required.

Reply:
More precisely, for this operation **floating point** input tensor `A[N_0, ..., N_r, 2]` is interpreted as a **complex** tensor `B = A[:, ... , : , 0] + iA[:, ... , : , 1]`, where `i` is an imaginary unit, and negative axes are **negative axes for the tensor** `B`.